File: algo/merge_sort.py
def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    
    mid = len(arr) // 2
    left_half = merge_sort(arr[:mid])
    right_half = merge_sort(arr[mid:])

    return merge(left_half, right_half)

def merge(left, right):
    sorted_array = []
    left_index = 0
    right_index = 0

    while left_index < len(left) and right_index < len(right):
        if left[left_index] < right[right_index]:
            sorted_array.append(left[left_index])
            left_index += 1
        else:
            sorted_array.append(right[right_index])
            right_index += 1

    while left_index < len(left):
        sorted_array.append(left[left_index])
        left_index += 1

    while right_index < len(right):
        sorted_array.append(right[right_index])
        right_index += 1

    return sorted_array

def test_merge_sort():
    arr = [1, 3, 2, 4, 5]
    arr = merge_sort(arr)
    assert arr == [1, 2, 3, 4, 5]

    arr = [5, 4, 3, 2, 1]
    arr = merge_sort(arr)
    assert arr == [1, 2, 3, 4, 5]

    arr = [1, 2, 3, 4, 5]
    arr = merge_sort(arr)
    assert arr == [1, 2, 3, 4, 5]

    arr = [1]
    arr = merge_sort(arr)
    assert arr == [1]

    arr = []
    arr = merge_sort(arr)
    assert arr == []

    print("All tests pass")

test_merge_sort()

File: algo/merge_sort.js
function mergeSort(arr){
    let length = arr.length;
    if(length <= 1){
        return arr;
    }

    let mid = Math.floor(length / 2);
    let left = mergeSort(arr.slice(0, mid));
    let right = mergeSort(arr.slice(mid, length));

    return merge(left, right);
};

function merge(left, right){
    let sortedArr = [];
    let leftIndex = 0;
    let rightIndex = 0;

    while(leftIndex < left.length && rightIndex < right.length){
        if(left[leftIndex] < right[rightIndex]){
            sortedArr.push(left[leftIndex]);
            leftIndex++;
        } else {
            sortedArr.push(right[rightIndex]);
            rightIndex++;
        }
    }

    while(leftIndex < left.length){
        sortedArr.push(left[leftIndex]);
        leftIndex++;
    }

    while(rightIndex < right.length){
        sortedArr.push(right[rightIndex]);
        rightIndex++;
    }

    return sortedArr;
}

function test_merge_sort(){
    const assert = require('assert');

    let arr = mergeSort([1, 4, 2, 3, 5]);

    assert(arr[0] === 1);
    assert(arr[1] === 2);
    assert(arr[2] === 3);
    assert(arr[3] === 4);
    assert(arr[4] === 5);

    console.log('All test cases pass');
}

test_merge_sort();